### PR TITLE
ControlLoop for controller manager

### LINKED ISSUE #726: Logic in ros2_control_node for timing
Here is my current understanding of how the logic in [ros2_control_node](https://github.com/ros-controls/ros2_control/blob/master/controller_manager/src/ros2_control_node.cpp) works:

ros2_control_loop drawio

There are several issues I see with this current design:
* The period value that is sent to read/update/write is always 0 because the `current_time` and `previous_time` values are always the same.
* The `time` value sent to read/update/write is the time between the call to `read` but before `update`.
* The `period` variable is shadowed inside the loop which makes it hard to understand at first where the value is coming from.
* This code is untestable because it is in a main function.

While the current design does not cause drift, the logic around these time and duration values is really confusing and hard to understand.

## Proposed changes

I propose to refactor the logic of this loop into a higher-order function (a function that takes functions as arguments) that can be tested. I will work out a diagram of how this new loop works and include unit tests of it.

Here is my proposed change to the loop structure:

ros2_control_loop_update drawio (1)

Here is some pseudocode of how it will work:

```cpp
template<typename NowF, typename OkF, typename WorkF, typename SleepF>
void ControlLoop(nanoseconds period, NowF now, OkF ok, WorkF do_work, SleepF sleep_until)
{
  auto current_time = now();
  auto measured_period = Duration{0};
  auto next_time = current_time;
  
  while(ok()) {
    auto const last_time = current_time;
    current_time = now();
    measured_period = current_time - last_time;
    next_time = next_time + period;

    do_work(current_time, measured_period);
    sleep_until(next_time);
  }
}
```

### PR DESCRIPTION
**Pull request by tylerjw**
Signed-off-by: Tyler Weaver <tylerjw@gmail.com>

Fixes #726

### CODE REVIEW

**Comment by tylerjw** [comment]
Please refer to the diagram and discussion in the related issue for reference to understanding this change.

File: controller_manager/src/ros2_control_node.cpp [hunk 1]
```diff
@@ -37,31 +44,33 @@ int main(int argc, char ** argv)
...
+      auto const now = [&cm]() { return cm->now(); };
+      auto const sleep_until = [](rclcpp::Time time)
+      {
+        std::this_thread::sleep_until(
+          std::chrono::system_clock::time_point(std::chrono::nanoseconds(time.nanoseconds())));
+      };
+      auto const ok = []() { return rclcpp::ok(); };
+      auto const do_work = [&cm](rclcpp::Time current_time, rclcpp::Duration period)
+      {
+        // Write is called first as the consistent rate or writing to hardware
```
Comment (gavanderhoorn) [inline controller_manager/src/ros2_control_node.cpp:65]:
```suggestion
        // Write is called first as the consistent rate of writing to hardware
```

File: controller_manager/test/test_control_loop.cpp [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+  };
+
+  // WHEN we run the ControlLoop
+  ControlLoop(10'000'000ns, now, ok, do_work, sleep_until);
+
+  // THEN we expect the second duration to be near the time the work function took
+  EXPECT_NEAR(durations.at(1).nanoseconds(), 20'000'000, 1'000'000);
+}
+
+TEST(TestControlLoop, LoopTimeBeforeWorkTime)
```
Comment (nbbrooks) [inline controller_manager/test/test_control_loop.cpp:128]:
```suggestion
TEST(TestControlLoop, WorkTimeAtOrAfterLoopTime)
```

File: controller_manager/test/test_control_loop.cpp [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+
+  // WHEN we run the ControlLoop
+  ControlLoop(10'000'000ns, now, ok, do_work, sleep_until);
+
+  // THEN we expect the work times to be after the loop times
+  EXPECT_GT(work_times.at(0), loop_times.at(0));
+  EXPECT_GT(work_times.at(1), loop_times.at(1));
+}
+
+TEST(TestControlLoop, LoopTimeSeparatedByPeriod)
```
Comment (nbbrooks) [inline controller_manager/test/test_control_loop.cpp:152]: How is this different than `DurationIsPeriod` test?
- manually calculates duration
- checks 1 loop rather than averaging over 10

bit of a nit, but my pref would to put those checks in the DurationIsPeriod test

File: controller_manager/test/test_control_loop.cpp [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+TEST(TestControlLoop, LoopTimeCloseToReportedDuration)
+{
+  auto loop_iterations = 0;
+  auto ok = [&loop_iterations]() {
+    loop_iterations++;
+    return loop_iterations < 3;
+  };
+
+  // GIVEN do_work function takes longer than the period
```
Comment (nbbrooks) [inline controller_manager/test/test_control_loop.cpp:180]:
```suggestion
  // GIVEN do_work function takes less time than the period
```

File: controller_manager/test/test_control_loop.cpp (hunk 3 above)
Comment (nbbrooks) [reply controller_manager/test/test_control_loop.cpp:152]: 'LoopTimeCloseToReportedDuration' seems to do this, but more explicitly

File: controller_manager/test/test_control_loop.cpp [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+  };
+
+  // WHEN we run the ControlLoop
+  ControlLoop(10'000'000ns, now, ok, do_work, sleep_until);
+
+  // THEN we expect the first duration to be near 0
+  EXPECT_NEAR(durations.at(0).nanoseconds(), 0, 1'000);
+}
+
+TEST(TestControlLoop, DoWorkDelayInDuration)
```
Comment (nbbrooks) [inline controller_manager/test/test_control_loop.cpp:106]:
```suggestion
TEST(TestControlLoop, DoWorkTimeGreaterThanPeriod)
```

nit - I think this would be clearer

File: controller_manager/test/test_control_loop.cpp (hunk 3 above)
Comment (nbbrooks) [reply controller_manager/test/test_control_loop.cpp:152]:
```suggestion
TEST(TestControlLoop, ReportedTimesMatchPeriod)
```
I see the differences now and made some naming suggestions from this fresh set of eyes

File: controller_manager/test/test_control_loop.cpp [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+
+  // WHEN we run the ControlLoop
+  ControlLoop(100'000'000ns, now, ok, do_work, sleep_until);
+
+  // THEN we expect the time between the loop times to be near the period
+  auto const duration = times.at(1) - times.at(0);
+  EXPECT_NEAR(duration.nanoseconds(), 100'000'000, 1'000'000);
+}
+
+TEST(TestControlLoop, LoopTimeCloseToReportedDuration)
```
Comment (nbbrooks) [inline controller_manager/test/test_control_loop.cpp:172]:
```suggestion
TEST(TestControlLoop, ReportedTimesMatchReportedDuration)
```

File: controller_manager/test/test_control_loop.cpp [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+  auto ok = []() { return false; };
+
+  // WHEN we run the ControlLoop
+  ControlLoop(1000ns, now, ok, do_work, sleep_until);
+
+  // THEN we expect do_work was never called, so the work_count should be 0
+  EXPECT_EQ(work_count, 0);
+}
+
+TEST(TestControlLoop, DurationIsPeriod)
```
Comment (nbbrooks) [inline controller_manager/test/test_control_loop.cpp:60]:
```suggestion
TEST(TestControlLoop, AverageReportedDurationMatchesPeriod)
```

**Review by nbbrooks — COMMENTED** [review]
I looked through the tests and made some suggestions around naming.

File: controller_manager/test/test_control_loop.cpp [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+  std::vector<rclcpp::Duration> durations;
+  auto do_work = [&durations](rclcpp::Time, rclcpp::Duration duration) {
+    durations.push_back(duration);
+  };
+
+  // WHEN we run the ControlLoop
+  ControlLoop(10'000'000ns, now, ok, do_work, sleep_until);
+
+  // THEN we expect the first duration to be near 0
+  EXPECT_NEAR(durations.at(0).nanoseconds(), 0, 1'000);
```
[the remark this answers is not in the document]
Comment (tylerjw) [reply controller_manager/test/test_control_loop.cpp:103]: I am happy to make this refactor. I was rushing to get this working before I left last week. I'll work on this PR now.

**Review by destogl — CHANGES_REQUESTED** [review]
We actually never planned for this control loop to grow so much, but this makes sense for sure! Thanks for the good explanations and for adding docs diagrams. Could you put those into the docs folder? And reference it so it does not get lost.

In general, I find the changes too complex only for the sake of testing. Is it possible to do this simpler?

File: controller_manager/src/ros2_control_node.cpp [hunk 9]
```diff
@@ -18,11 +18,18 @@
 #include <string>
 #include <thread>
 
+#include "controller_manager/control_loop.hpp"
 #include "controller_manager/controller_manager.hpp"
 #include "rclcpp/rclcpp.hpp"
 
 using namespace std::chrono_literals;
 
+void sleep_until(rclcpp::Time time)
```
[the remark this answers is not in the document]
Comment (tylerjw) [reply controller_manager/src/ros2_control_node.cpp:27]: It is because it enables testing this functionality without ever putting a sleep in the test. You should read the tests to understand why this is useful.

**Comment by tylerjw** [comment]
> In general, I find the changes too complex only for the sake of testing. Is it possible to do this simpler?

This is taking a generic programming approach to solving this problem. I generally think correctness and robust testing to be much more valuable than reducing the number of things the programmer has to learn to read the code.

I can find you references if you'd like for why I am using template arguments over using `std::function` or something similar. The short answer is that template arguments here will generate much more performant code because `std::function` has to do heap allocations to enable its copy constructor and allow you to keep collections of them.

I do think this function is complex enough and separate enough from the main function to justify it being in its own file. If this was trivially easy to write in a correct way why was the main loop itself so broken? In my experience, code that is not designed to be tested is almost always trivially broken because it is too hard to program without errors.

File: controller_manager/include/controller_manager/control_loop.hpp [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+ * \param[in]  now invocable that returns rclcpp::Time
+ * \param[in]  ok invocable that returns true if the loop should continue
+ * \param[in]  do_work invocable that takes the current time and measured period
+ *             and does the work of the control loop
+ * \param[in]  sleep_until invocable that takes a rclcpp::Time and sleeps until
+ *             that time
+ */
+template <
+  typename NowInvocable, typename OkInvocable, typename WorkInvocable, typename SleepInvocable>
+void ControlLoop(
```
[the remark this answers is not in the document]
Comment (tylerjw) [reply controller_manager/include/controller_manager/control_loop.hpp:34]: It would not be testable if put in the `.cpp` file because I need to create a new version of it in the translation unit that is the test. This is the reason that templates always have to go in a header file if you want to use them in two separate translation units.

File: controller_manager/src/ros2_control_node.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (tylerjw) [reply controller_manager/src/ros2_control_node.cpp:27]: You are right that it isn't needed as I also wrote this as a capture below :facepalm:

**Comment by tylerjw** [comment]
> Could you put those into the docs folder? And reference it so it does not get lost.

I added an explanation of the control loop to the userdoc.rst and included the diagram.

> In general, I find the changes too complex only for the sake of testing. Is it possible to do this simpler?

Is there something specific you think is too complex? I do not understand what about this change is more complex than what is normal in this project.

**Comment by tylerjw** [comment]
If you are curious about why I chose to use template arguments for the function parameters here are a couple of links. The basic reason is that because I don't need to store the functions in a runtime collection I use templates. In general, you should always use templates when passing functions if you can because if the type is known at compile time your compiler can do a much better job of optimizing it. If you are curious how std::function works and allows us to store function objects at runtime without inheritance, it uses a technique called type-errasure. The downside to type-erasure is that it has to do a heap allocation and does use virtual function dispatch via inheritance of an internal model template. This trick is a way of having a duck-typed interface in C++ without inheritance.
- https://rules.sonarsource.com/cpp/tag/bad-practice/RSPEC-5213
- https://stackoverflow.com/a/14678298

File: controller_manager/include/controller_manager/control_loop.hpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (tylerjw) [reply controller_manager/include/controller_manager/control_loop.hpp:36]: I added static_assert statements to this to make the compile errors nicer if a user tries to use this function and makes a mistake in the sort of function-like objects they pass into this.

[the remark this answers is not in the document]
Comment (tylerjw) [reply controller_manager/include/controller_manager/control_loop.hpp:35]: > Furthermore, do we really need methods as arguments? IMHO, for this simple functionality this looks very overengineered.

Do you have a suggestion for how to make this behavior testable without methods as arguments? IMHO, code that is not designed to be tested is almost always broken.

[the remark this answers is not in the document]
Comment (gavanderhoorn) [reply controller_manager/include/controller_manager/control_loop.hpp:35]: > code that is not designed to be tested is almost always broken.

bit heavy-handed, isn't it?

I agree testability is good, but you cannot generalise like you're doing here. It isn't very constructive.

[the remark this answers is not in the document]
Comment (tylerjw) [reply controller_manager/include/controller_manager/control_loop.hpp:35]: I find it is much harder to make mistakes myself when I write tests. I find writing correct code without tests is often more difficult than writing code with tests. I find that writing code that can be tested, while always more work than the untestable variant, is less work for others to use... As it is less likely someone will need to correct my mistakes later.

Obviously writing tests doesn't eliminate all mistakes, it just makes them more difficult to create.

Statements about tests here may contain hyperbole.

[the remark this answers is not in the document]
Comment (tylerjw) [reply controller_manager/include/controller_manager/control_loop.hpp:35]: There is plenty of working code without tests. If the maintainers of this project would rather this be fixed without tests with a smaller PR I can trivially provide that PR. Much of the value of these tests is that they helped me reason about how to write this and what the corner cases of it is.

[the remark this answers is not in the document]
Comment (tylerjw) [reply controller_manager/include/controller_manager/control_loop.hpp:35]: I'm sorry for being sarcastic and mean here. Please let me know what feels overly complex or over engineered. I'm happy to make changes or add documentation to make this code more readable or better in any way.

[the remark this answers is not in the document]
Comment (gavanderhoorn) [reply controller_manager/include/controller_manager/control_loop.hpp:35]: There is no need to apologise Tyler.

Obviously introducing tests is better than not doing it, and adapting code structure to allow for testing is better than not doing it.

There are different ways of testing though, and unit tests are only one of them. Not all of them require dependency injection or mocking.

[the remark this answers is not in the document]
Comment (tylerjw) [reply controller_manager/include/controller_manager/control_loop.hpp:35]: I tried testing this loop externally with an integration test and was unable to figure out how to do it robustly or simply. My current robot driver I'm working on is effectively a test of this but I do think it is way too much code for a test.

[the remark this answers is not in the document]
Comment (gavanderhoorn) [reply controller_manager/include/controller_manager/control_loop.hpp:35]: You're misunderstanding me: I'm not telling you this is not a good approach, or you should use a different testing strategy.

I'm saying asserting that only "testable/tested code is fault-free" is not a claim you can make, and that using a different approach than dependency injection/mocking could also be used in some cases.

I'm not saying in *this* case.

[the remark this answers is not in the document]
Comment (tylerjw) [reply controller_manager/include/controller_manager/control_loop.hpp:35]: "testable/tested code is fault-free" is an exaggeration for sure. As a person who makes a lot of mistakes, and finds writing code that does what I want hard, I find writing tests useful.

Theoretically, tests also help you in the future when you come back to something and want to make a change but not break what already exists.

**Comment by tylerjw** [comment]
Closing as this is a dead PR.
